tools/zetasql-helper/zetasql-toolkit-core/src/main/java/com/google/zetasql/toolkit/catalog/bigquery/BigQueryAPIResourceProvider.java

```diff
@@ -0,0 +1,660 @@
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.MaterializedViewDefinition;
+import com.google.cloud.bigquery.Routine;
+import com.google.cloud.bigquery.RoutineArgument;
+import com.google.cloud.bigquery.RoutineId;
+import com.google.cloud.bigquery.SnapshotTableDefinition;
+import com.google.cloud.bigquery.StandardSQLDataType;
+import com.google.cloud.bigquery.StandardSQLStructType;
+import com.google.cloud.bigquery.StandardSQLTableType;
+import com.google.cloud.bigquery.StandardSQLTypeName;
+import com.google.cloud.bigquery.StandardTableDefinition;
+import com.google.cloud.bigquery.Table;
+import com.google.cloud.bigquery.TableDefinition;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TimePartitioning;
+import com.google.common.collect.ImmutableList;
+import com.google.zetasql.Function;
+import com.google.zetasql.FunctionArgumentType;
+import com.google.zetasql.FunctionArgumentType.FunctionArgumentTypeOptions;
+import com.google.zetasql.FunctionSignature;
+import com.google.zetasql.SimpleColumn;
+import com.google.zetasql.SimpleTable;
+import com.google.zetasql.StructType.StructField;
+import com.google.zetasql.TVFRelation;
+import com.google.zetasql.Type;
+import com.google.zetasql.TypeFactory;
+import com.google.zetasql.ZetaSQLFunctions.FunctionEnums.Mode;
+import com.google.zetasql.ZetaSQLFunctions.FunctionEnums.ProcedureArgumentMode;
+import com.google.zetasql.ZetaSQLFunctions.SignatureArgumentKind;
+import com.google.zetasql.ZetaSQLType.TypeKind;
+import com.google.zetasql.toolkit.catalog.CatalogOperations;
+import com.google.zetasql.toolkit.catalog.bigquery.BigQueryService.Result;
+import com.google.zetasql.toolkit.catalog.bigquery.exceptions.BigQueryAPIError;
+import com.google.zetasql.toolkit.catalog.bigquery.exceptions.InvalidBigQueryReference;
+import com.google.zetasql.toolkit.catalog.bigquery.exceptions.MissingRoutineReturnType;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * {@link BigQueryResourceProvider} implementation that uses the BigQuery API to get the BigQuery
+ * resources. Resources are cached internally, so multiple request to the same resource only hit the
+ * API once.
+ */
+public class BigQueryAPIResourceProvider implements BigQueryResourceProvider {
+
+  private final BigQueryService service;
+
+  /** Constructs a BigQueryAPIResourceProvider that uses the default BigQuery client */
+  public BigQueryAPIResourceProvider() {
+    this(BigQueryOptions.newBuilder().build().getService());
+  }
+
+  /**
+   * Constructs a BigQueryAPIResourceProvider that uses the provided BigQuery client.
+   *
+   * @param client The BigQuery client this instance should use
+   */
+  public BigQueryAPIResourceProvider(BigQuery client) {
+    this(new BigQueryService(client));
```

Comment:
**High priority!**

Please set headerProvider

Reply:
Added PSO code tracking:

* for BigQuery and Spanner API calls on commit https://github.com/GoogleCloudPlatform/professional-services/pull/1005/commits/10faf86a54e103649bb9a6d810682ad529a0a703.
* for analysis and validation usage on commit https://github.com/GoogleCloudPlatform/professional-services/pull/1005/commits/3feccbe6eed1e88ac001fba87dffed1452ef42cd